Feature: Enable access controlled links

NinaHerrmann opened the pull request:
WIP!
This pull request aims to insert the function of a controlled link/access controlled link. This link ensures that after submission files can merely restrictedly changed and the access of the file is controlled with private links rather than public links. 
The Google drive and OneDrive repositories are examples where the functionality has been implemented.

Currently the functionalities are not sufficiently tested.

NinaHerrmann committed:
updated lib.php supported_returntypes now allows controlled link when systemaccount is connected

NinaHerrmann committed:
Removed setting header in callback function, since this is done in the patch of MDL-59512

NinaHerrmann committed:
Updated lib.php restored setHeader, improved documentation

NinaHerrmann committed:
updated test for supported filetypes

NinaHerrmann committed:
updated lib.php and ocs_client. The work is still in progress. A share is generated and in the systemaccount a path for the file is created. However, there are a lot of todos. NEXT copy the file to the created path

NinaHerrmann committed:
updated the reference_file_selected function.

NinaHerrmann committed:
inserted a basic test for refernce_file_selected

NinaHerrmann committed:
Inserted additional test for private functions used in reference_file_selected

NinaHerrmann committed:
WIP: access controlled links can be created, and when a authenticated users clicks the file a private share is established. However, there exist several cases when the process failed therefore WIP.

NinaHerrmann committed:
WIP updating the lib.php to merge it with the updated version of the owncloud_client.php 1 step of creating a access controlled link works again (creating a share between user and technical user) NEXT: create folders

NinaHerrmann committed:
WIP updating the lib.php to merge it with the updated version of the webdav_client.php step two and three are working again (creating a path in the systemaccount and moving the file to the path) NEXT display file to user in moodle

NinaHerrmann committed:
inserted FILE_EXTERNAL again as supported returntype

NinaHerrmann committed:
updated is_empty check

NinaHerrmann committed:
updated .travis.yml

NinaHerrmann committed:
updated way to generate webdavprefix

NinaHerrmann committed:
inserted double check for sysauth connected

NinaHerrmann committed:
insert 403 as valid option (share was already created)

NinaHerrmann committed:
deleted method to create private links not longer needed

NinaHerrmann committed:
added function to return type option names

NinaHerrmann committed:
updated default returntype

NinaHerrmann committed:
updated travis.yml

NinaHerrmann committed:
generates a folder called 'Moodlefiles" to store Moodlefiles collectively

NinaHerrmann committed:
updated send_file moves file to folder

NinaHerrmann commented:
To increase clarity, the essential tasks to be done before the core functionality is usable are listed:

- [x] **Pop-up window for authentication**
- [x] **File is moved as a share in the users account** 
- [x] **Settings-field for the name of the folder in owncloud.** The default value is _Moodlefiles_.
- [x] **Settings-field for the time a file should be shared.** The default value is 1 week.
- [x] **Handle moving a file when file is renamed in case file with same name already exists.** when file is moved the information about the share is requested which includes possible renaming.
- [x] **Redirection url is built with the file-id.** Webdav redirect enforces download.
- [x] **Redirect to download** Done with a header to the webdav-dowload-url.
- [x] **Display different file types** Redirecting with webdav has the major benefit that pictures are displayed embedded. 
- [x] **Handle multiple files** Not considered as a severe problem, since normal files do not offer that functionality.

NinaHerrmann committed:
updated send_file in case a user is not authenticated he/she is directly redirected to login page

NinaHerrmann committed:
Additional comment to the last commit when redirecting from ownCloud to moodle the following error is displayed : If parent window is on HTTPS, then we are not allowed to access window.opener object, so we cannot refresh the repository for you automatically, but we already got your session, just go back to file picker and select the repository again, it should work now. However, the redirect does not work, therefore testing whether this works with two https sides is necessary

NinaHerrmann committed:
updated send_file() now redirects directly to the webdav url and therefore file is immediately downloaded. However, name of folder is hardcoded to Moodlefiles

NinaHerrmann committed:
updated send_file() moves share (!) to the folder Moodlefiles :tada:

NinaHerrmann committed:
updated travis.yml force node version

NinaHerrmann committed:
updated lib.php send_file link is generated with fileid TODO: if share already created get fileid

NinaHerrmann committed:
updated lib.php send_file() redirection to link with file id when file was already shared

NinaHerrmann committed:
corrected documentation in ocs_client

NinaHerrmann committed:
updated lib.php send_file() when user is not authenticated pop up appears and redirect works

NinaHerrmann committed:
updated default time for share to 1 week

NinaHerrmann committed:
WIP adding a setting to determine time a file is shared

NinaHerrmann committed:
updated lib.php settings for a timeinterval to create a share

NinaHerrmann committed:
deleted old todo

NinaHerrmann committed:
updated lib.php setting added field for custom folder name

NinaHerrmann committed:
updated lib.php config setting for foldername included in code.

NinaHerrmann committed:
updated lib.php settings form added a warning for changing the foldername for access controlled links

NinaHerrmann committed:
updated lib.php use shareid and information about share to backtrace possible changes in filename

NinaHerrmann committed:
updated lib.php inserted new todos and deleted old

NinaHerrmann committed:
updated lib.php added foreach element for get_shares

NinaHerrmann committed:
updated lib.php inserted foreach for get information about share xml answer and inserted webdav redirect (enforces download.)

NinaHerrmann committed:
updated lib.php removed minor violations recommended by travis

NinaHerrmann committed:
updated phpunit test fixed old test to new expected values and inserted additional options in initialisation

NinaHerrmann committed:
inserted new file for test for access controlled links

NinaHerrmann committed:
updated test for access controlled links added test for create folder path in case of success

NinaHerrmann committed:
updated lib.php and test is_dir is now checked more restrictively, inserted test for folders have to be created. This need to be tested in a separate function since phpunit do not allow to overwrite return values for mocks in the same function

NinaHerrmann committed:
updated tests modularized create_folder_path

NinaHerrmann committed:
updated lib.php check for folder created more restrictively

NinaHerrmann committed:
inerted test for copy file to path. Assures that webdav function is called with the right params

NinaHerrmann committed:
updated lib.php fixed bug when share is created filetarget has to be checked twice

NinaHerrmann committed:
updated lib.php basic test for private delete share function

NinaHerrmann committed:
updated lib.php stored systemocsclient as global variable

NinaHerrmann committed:
deleted old copy_file_to_path testing function

NinaHerrmann committed:
inserted function to test when systemaccount creates share and renamed function for clarity

NinaHerrmann committed:
extended test reference_file_selected function

NinaHerrmann committed:
updated lib.php minor adjustments

NinaHerrmann committed:
inserted start of test function for send_file

NinaHerrmann committed:
updated the error message

NinaHerrmann committed:
inserted additional part for testing send_file

NinaHerrmann committed:
changed redirect in send_file now a success message is displayed on the next page requested. However, alternatives should be discussed since it might confuse users.

NinaHerrmann commented:
The todos for the next commits:
- [x] **Discuss how errors should be displayed** (e.g. When file can no longer be accessed) possible solutions: email admin, show message to user, contact course admin? (see .pdf files)
- [x] **Insert basic phpunit test for all functions and fix existing test.** 

- [ ] **Users can not differentiate between _normal_ files and _access_controlled links_.** This is disturbing for the user, since a file should merely be shared with the user and downloaded when he/she knowns about the procedure. This is partly handled by printing an information message, however, this is merely displayed on the next Moodle page loaded.
- [ ] **Handle login in case of images** Images are being displayed on the Moodle page. However, this suppresses the login button (and the corresponding pop-up) in case the user is not authenticated. Therefore, no content is displayed. The google repository has the same problem. I see three ways to solve this issue:

1. Firstly, the display type of the resource could be changed from `RESOURCELIB_DISPLAY_AUTO` (which transfers to `RESOURCELIB_DISPLAY_EMBED` look at `mod/resource/view.php`) to e.g. `RESOURCELIB_DISPLAY_DOWNLOAD` or `RESOURCELIB_DISPLAY_FRAME`. Therefore, the corresponding database entry needs to be updated in the function `reference_file_selected()`. It might be troublesome to get the id for updating the file.
2. Secondly, `send_file()` could be adjusted to behave differently for pictures. However, I am unsure to which extent the function can still change the side. 
3. Thirdly, this issue could be handled by changing the Moodle core. This seems most reasonable since Google encounters the same problem in their repository. 

- [ ] **Handle log out system account** This really distress me since for unknown reason the system account is sometimes unable to connect and is merely able to update the token in case the administrator logs in again. It seems to appear when shares are manually deleted. However, this has still to be proven. It might require additionally some research in the ownCloud code.
- [ ] **Research default storage** There might be a default location for ownCloud users to store their downloads. In this case the Moodlefolder should be embedded into this default location.

NinaHerrmann committed:
minor changes in throwing errors. Reactions should still be discussed collaboratively.

NinaHerrmann committed:
improved error throwing of reference_file selected and minor code style fixes

NinaHerrmann committed:
minor codestyle fixes in tests

NinaHerrmann committed:
updated language string for exception handling

NinaHerrmann committed:
updated lib.php throwing of exceptions

NinaHerrmann committed:
updated test for exception handling of send_file

NinaHerrmann commented:
For easy reference to error logging
reference_file_selected:
[reference_file_selected.pdf](https://github.com/learnweb/moodle-repository_owncloud/files/1562207/reference_file_selected.pdf)
send_file:
[send_file.pdf](https://github.com/learnweb/moodle-repository_owncloud/files/1562208/send_file.pdf)

NinaHerrmann committed:
outsourced part of the functions of reference_file_selected into a separate class

NinaHerrmann commented:
FYI: outsourcing parts of the access_controlled_link functions is still WIP.

NinaHerrmann committed:
started inserting test for new class access_controlled_link_manager

NinaHerrmann committed:
amendment forgot to delete unnecessary testing line

NinaHerrmann committed:
outsourced functions of lib.php send_file() to the access_controlled_link_manager class and starting cleanup

NinaHerrmann committed:
deleted outdated test TODO: writhe new ones for class

NinaHerrmann committed:
delted outdated test lib.php

NinaHerrmann committed:
updated access_controlled_link_manager class

NinaHerrmann committed:
added @group annotation for ocs_test

NinaHerrmann committed:
trying to set a expirationdate for shares however does not work yet

NinaHerrmann committed:
after wondering why the expirationDate param does not work found out that owncloud merely support that feature for public links. However, this feature might be available in the future, see :https://github.com/owncloud/android/issues/1973

NinaHerrmann committed:
inserted an additional group annotation and inserted test for transfer_file_to_path()

NinaHerrmann committed:
inserted explanatory comment on expected exception for contrusctor of linkmanager in testcase

NinaHerrmann committed:
inserted test for linkmanager class for function create_folder_path_folders_are_not_created()

NinaHerrmann committed:
inserted test for linkmanager for method create_system_dav and minor code style fixes

NinaHerrmann committed:
added test for linkmanager for method create_storage_folder (two methods are still missing in the tests)

NinaHerrmann committed:
added additional test class for access_controlled_link_manager with different constructor.

NinaHerrmann committed:
updated access_controlled_link_manager class set properties to peotected and minor code fixes

NinaHerrmann committed:
updated the testclass for the access_controlled_link manager. The linkmanager is now created in a testclass which inherits from the original class.

NinaHerrmann committed:
nearly forgot to change the year but we want to be thoroughly here

NinaHerrmann committed:
updated access_controlled_link class and tests: inserted a new variable in the class to save the ocsoauthclient and added a test for get_shares_from_path()

NinaHerrmann committed:
updated tests for accesscontrolled link manager inserted test for create_system_dav()

NinaHerrmann committed:
updated test for access_controlled_link manager, inserted try catch block to catch error in case a \core\invalid_persistent_exception is thrown (http is not allowed for endpoints, however the exception is not always thrown) and added test for function get_share_information_from_shareid()

Dagefoerde committed:
add convenience cli script

Dagefoerde commented:
warning to other developers: rebased onto current `master`.

Dagefoerde commented:
Does not work. I tried to add an access controlled link and then submit the filepicker.
Screenshot (Moodle) after submitting the filepicker: 
![clicking on save](https://user-images.githubusercontent.com/432117/40661737-cfe3f6e0-6354-11e8-88cb-b6e63c186448.png)

ownCloud log output:
```
[Tue May 29 15:20:36 2018] 172.17.0.10:38464 [207]: /remote.php/webdav                                                                                                                                             
[Tue May 29 15:20:36 2018] 172.17.0.10:38466 [201]: /remote.php/webdav//System                                                                                                                                     
[Tue May 29 15:20:36 2018] 172.17.0.10:38468 [207]: /remote.php/webdav//System                                                                                                                                     
[Tue May 29 15:20:37 2018] 172.17.0.10:38470 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde                                                                                                         
[Tue May 29 15:20:37 2018] 172.17.0.10:38472 [207]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde                                                                                                         
[Tue May 29 15:20:37 2018] 172.17.0.10:38474 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user                                                                                                    
[Tue May 29 15:20:37 2018] 172.17.0.10:38476 [207]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user                                                                                                    
[Tue May 29 15:20:37 2018] 172.17.0.10:38478 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user/private                                                                                            
[Tue May 29 15:20:37 2018] 172.17.0.10:38480 [207]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user/private                                                                                            
[Tue May 29 15:20:37 2018] 172.17.0.10:38482 [201]: /remote.php/webdav//System/User%20Jan%20Dagef%C3%B6rde/user/private/0                                                                                          
[Tue May 29 15:20:42 2018] Exception: {"Exception":"Error","Message":"Maximum function nesting level of '256' reached, aborting!","Code":0,"Trace":"#0 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(1149): Doctrine\\DBAL\\Query\\QueryBuilder->getSQLForJoins('`*PREFIX*extern...', Array)
#1 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(1112): Doctrine\\DBAL\\Query\\QueryBuilder->getFromClauses()
#2 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(244): Doctrine\\DBAL\\Query\\QueryBuilder->getSQLForSelect()
#3 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php(206): Doctrine\\DBAL\\Query\\QueryBuilder->getSQL()
#4 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/DB\/QueryBuilder\/QueryBuilder.php(141): Doctrine\\DBAL\\Query\\QueryBuilder->execute()
#5 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/DBConfigService.php(337): OC\\DB\\QueryBuilder\\QueryBuilder->execute()
#6 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/DBConfigService.php(103): OC\\Files\\External\\Service\\DBConfigService->getMountsFromQuery(Object(OC\\DB\\QueryBuilder\\QueryBuilder))
#7 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(73): OC\\Files\\External\\Service\\DBConfigService->getAdminMounts()
#8 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(129): OC\\Files\\External\\Service\\StoragesService->readDBConfig()
#9 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(179): OC\\Files\\External\\Service\\StoragesService->readConfig()
#10 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/Service\/StoragesService.php(188): OC\\Files\\External\\Service\\StoragesService->getAllStorages()
#11 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/External\/LegacyUtil.php(100): OC\\Files\\External\\Service\\StoragesService->getStorages()
#12 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Encryption\/Util.php(303): OC\\Files\\External\\LegacyUtil::getSystemMountPoints()
#13 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Encryption\/Keys\/Storage.php(343): OC\\Encryption\\Util->isSystemWideMountPoint('\/files\/System\/U...', 'admin')
#14 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Encryption\/Keys\/Storage.php(323): OC\\Encryption\\Keys\\Storage->getPathToKeys('\/admin\/\/files\/S...')
#15 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(1030): OC\\Encryption\\Keys\\Storage->copyKeys('\/admin\/\/files\/S...', '\/admin\/files\/Sy...')
#16 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(770): OC\\Files\\Storage\\Wrapper\\Encryption->copyKeys('\/admin\/\/files\/S...', '\/admin\/files\/Sy...')
#17 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#18 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#19 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#20 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#21 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#22 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#23 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#24 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#25 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#26 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#27 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#28 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#29 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#30 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#31 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#32 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
<snip>
#229 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#230 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#231 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#232 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#233 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#234 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#235 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#236 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#237 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#238 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System\/Us...', 'files\/System\/Us...', false, false)
#239 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System', 'files\/System\/Us...', false, false)
#240 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(781): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files\/System', 'files\/System\/Us...', false, false)
#241 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(679): OC\\Files\\Storage\\Wrapper\\Encryption->copyBetweenStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files', 'files\/System\/Us...', false, false)
#242 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Encryption.php(358): OC\\Files\\Storage\\Wrapper\\Encryption->copyFromStorage(Object(OC\\Files\\Storage\\Wrapper\\Encryption), 'files', 'files\/System\/Us...')
#243 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php(283): OC\\Files\\Storage\\Wrapper\\Encryption->copy('files', 'files\/System\/Us...')
#244 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php(283): OC\\Files\\Storage\\Wrapper\\Wrapper->copy('files', 'files\/System\/Us...')
#245 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/View.php(945): OC\\Files\\Storage\\Wrapper\\Wrapper->copy('files', 'files\/System\/Us...')
#246 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/public\/Events\/EventEmitterTrait.php(50): OC\\Files\\View->OC\\Files\\{closure}(Array)
#247 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/private\/Files\/View.php(988): OC\\Files\\View->emittingCall(Object(Closure), Array, 'file', 'copy')
#248 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/apps\/dav\/lib\/Connector\/Sabre\/ObjectTree.php(250): OC\\Files\\View->copy('\/', '\/System\/User Ja...')
#249 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/CorePlugin.php(684): OCA\\DAV\\Connector\\Sabre\\ObjectTree->copy('', 'System\/User Jan...')
#250 [internal function]: Sabre\\DAV\\CorePlugin->httpCopy(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))
#251 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/event\/lib\/EventEmitterTrait.php(105): call_user_func_array(Array, Array)
#252 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(479): Sabre\\Event\\EventEmitter->emit('method:COPY', Array)
#253 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/sabre\/dav\/lib\/DAV\/Server.php(254): Sabre\\DAV\\Server->invokeMethod(Object(Sabre\\HTTP\\Request), Object(Sabre\\HTTP\\Response))
#254 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/apps\/dav\/appinfo\/v1\/webdav.php(63): Sabre\\DAV\\Server->exec()
#255 \/home\/j_dage01\/Development\/Learnweb\/owncloud\/remote.php(165): require_once('\/home\/j_dage01\/...')
#256 {main}","File":"\/home\/j_dage01\/Development\/Learnweb\/owncloud\/lib\/composer\/doctrine\/dbal\/lib\/Doctrine\/DBAL\/Query\/QueryBuilder.php","Line":1306}
[Tue May 29 15:20:42 2018] 172.17.0.10:38484 [500]: /remote.php/webdav/
```

NinaHerrmann committed:
doing some cleanup on the way

NinaHerrmann committed:
changed the way the ocs_client build the url for request that require an id

NinaHerrmann committed:
adjusted README.md

NinaHerrmann committed:
adjusted constructor of ocs_client

Dagefoerde commented on classes/ocs_client.php:
```diff
@@ -41,56 +41,113 @@ class ocs_client extends rest {
         return [
             'create_share' => [
-                'endpoint' => $this->ocsendpoint,
+                'endpoint' => $this->ocsendpoint->out(),
```
`->out(false)`

Dagefoerde commented:
- [ ]  Files are shared with no restrictions regarding the usage; i.e. users who gain access to an access controlled link are allowed to re-share and to edit(!) the file. I have not compared this to Moodle+Google Docs et al, but I guess this is not intended behaviour.

Dagefoerde committed:
OCS: Share URLs must not be escaped as they are not used in output

Dagefoerde committed:
lib: Fix order of checks to $issuer

Dagefoerde committed:
if logged in, do not continue blindly!

Before, this tried to create a folder for the (not so) logged in user,
which resulted in an exception

Dagefoerde committed:
Merge branch 'master' into NinaHerrmann-feature/file_conrolled_link

Dagefoerde committed:
ocs: fixed duplicate const declaration

Dagefoerde committed:
tests: Fix mock (exactly(2) only happens on test systems)

Dagefoerde commented:
- [x] Folder "Moodlefiles" displays a warning that it should never be changed. I don't think this is correct, because the behaviour is actually very resilient to changing foldernames: If a share already exists, you get the correct location via the OCS API, regardless of the used folder (users may even move around the file without issue). If it doesn't, the file is shared into the (then active) folder, which is perfect. Worst case scenario is that users accumulate a bunch of "Moodlefiles" folders because the admins change the name frequently... But that shouldn't be a reason for presenting this warning.
- [ ] On accessing a file, if the user is not logged in in ownCloud, a button is shown that leads the user to the login. In some cases, the user can't see this button, because a) it is in an invisible iframe, b) the file is referenced via `<img src=`. We cannot do much about b), ~~but we can circumvent a) by showing a popup right away (skipping the button).~~ and not much about a) either. --- Tried that, doesn't work: The popup (obviously) gets blocked by a popup blocker :( attempt is in `Dagefoerde/popup-on-sendfile`.

Dagefoerde committed:
Settings: Remove warning

Dagefoerde commented:
- [x] Access controlled shares never expire – Admins can set an expiration date which is (probably) transmitted to oC/Nc on creation of a share. However, both don't support expiration of private shares, so that setting is effectively useless. [ Update: ] The setting is not used at all and can be safely deleted.
- [x] Google and OneDrive repositories allow settings that control how files go into Moodle (option to select between only copying, access controlled links, or both)

Dagefoerde committed:
Remove timeintervalsharing option as it is neither supported nor used

Dagefoerde committed:
Settings: Group issuer setting and help text together again

Dagefoerde committed:
Settings: Add and use settings for allowed and default return types

Dagefoerde committed:
Settings: Add language strings for new settings

Dagefoerde commented:
- [x] Folder structure in system account heavily depends on users' chosen frontend languages, i.e. files might not be found because a wrong language is assumed (fix cf. https://github.com/moodle/moodle/commit/e0a903243be16b461cedfc223d59abe42bd527a2#diff-7066ab7ec56b9a7f65f10f907f4dc15a)

Dagefoerde committed:
File picker: Add button to go to web UI

Consistent with Google Drive repository

Dagefoerde committed:
Types: Reinstate external/reference types

For their meanings, refer to
https://docs.moodle.org/dev/Repository_plugins#supported_returntypes.28.29

Dagefoerde committed:
acl: Make sure language in system account's storage is always system
language

Also, add unique IDs based on context IDs.

Dagefoerde committed:
Simplify returned types and parameter types; re-add FILE_REFERENCE

Dagefoerde committed:
(phpcs fixes)

Dagefoerde committed:
bump version no for pre-release

Dagefoerde committed:
last-minute-fix: restore name property in references

it holds the original filename

Dagefoerde committed:
Upgrade step for converting old-style references to new JSON represent.

Dagefoerde committed:
Add scripts to travis builds

Dagefoerde committed:
Merge test cases

Actually, all these tests are lib tests. phpcpd was right to complain
about the shared setUp methods.

Dagefoerde committed:
Use require instead of require_once

as per results from current discussion in dev chat.

Dagefoerde committed:
Even without a system account, conventional links would still work

Dagefoerde committed:
Settings: Move default behaviour choice to the top

Because the first item is selected for those who upgraded from a
previous version and do not have that setting yet. "Both" is the
default behaviour for those who do not have that setting.

Dagefoerde commented:
- [x] access controlled shares from the system account to (student) users may not be writeable by users!

Dagefoerde committed:
Access controlled link: Restrict write/share access

Only those who may write to the course module should be able to write to
the resulting file. All the others only get read access!

Dagefoerde committed:
upgrade: when reference is changed, also change referencehash!

Dagefoerde commented:
Needs new docs, then it's ready for merging and for releasing a new version.

Dagefoerde committed:
Update documentation
